File: make_favicon.py
import sys
try:
    from PIL import Image, ImageDraw
except ImportError:
    import subprocess
    subprocess.check_call([sys.executable, "-m", "pip", "install", "Pillow"])
    from PIL import Image, ImageDraw

img_path = "/Users/tomhome/.gemini/antigravity/brain/ea2a3963-d5a6-4338-bd31-61f6c8318a40/tm_logo_purple_cream_1788249235599.jpg"
img = Image.open(img_path).convert("RGBA")

# Create a circular mask
mask = Image.new('L', img.size, 0)
draw = ImageDraw.Draw(mask)
# Find the bounding box of the purple circle.
gray = img.convert("L")
bw = gray.point(lambda x: 0 if x > 240 else 255, '1')
bbox = bw.getbbox()

if bbox:
    # Slightly shrink the bounding box to avoid white fringes
    b_margin = 2
    b = (bbox[0]+b_margin, bbox[1]+b_margin, bbox[2]-b_margin, bbox[3]-b_margin)
    draw.ellipse(b, fill=255)
else:
    # fallback
    draw.ellipse((50, 50, img.size[0]-50, img.size[1]-50), fill=255)

# apply mask
result = img.copy()
result.putalpha(mask)

# Save as icon.png in the app directory
result.save("/Users/tomhome/Development/tommares-portfolio/src/app/icon.png", "PNG")
# Overwrite the default favicon.ico
result.resize((256, 256)).save("/Users/tomhome/Development/tommares-portfolio/src/app/favicon.ico", format="ICO", sizes=[(256, 256)])
print("Favicon created!")

File: make_favicon.js
const sharp = require('sharp');
const path = require('path');

const imgPath = '/Users/tomhome/.gemini/antigravity/brain/ea2a3963-d5a6-4338-bd31-61f6c8318a40/tm_logo_purple_cream_1788249235599.jpg';
const destPng = path.join(__dirname, 'src', 'app', 'icon.png');

async function makeFavicon() {
  const { data, info } = await sharp(imgPath)
    .ensureAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const w = info.width;
  const h = info.height;
  
  // Find the exact bounds of the purple circle by checking the middle row
  const cy = Math.floor(h / 2);
  let left = 0;
  let right = w - 1;
  
  // A pixel is background if it's close to white (R>240, G>240, B>240)
  function isBg(x, y) {
    const idx = (y * w + x) * 4;
    return data[idx] > 240 && data[idx+1] > 240 && data[idx+2] > 240;
  }
  
  while (left < w / 2 && isBg(left, cy)) left++;
  while (right > w / 2 && isBg(right, cy)) right--;
  
  let top = 0;
  let bottom = h - 1;
  const cx = Math.floor(w / 2);
  while (top < h / 2 && isBg(cx, top)) top++;
  while (bottom > h / 2 && isBg(cx, bottom)) bottom--;
  
  const widthBox = right - left;
  const heightBox = bottom - top;
  const radius = Math.min(widthBox, heightBox) / 2 - 4; // -4 to remove any anti-aliased edge
  
  const realCx = left + widthBox / 2;
  const realCy = top + heightBox / 2;

  // We want the final image to be exactly the bounding box of the circle so there is no padding at all
  const cropSize = Math.floor(radius * 2);
  const extractLeft = Math.floor(realCx - radius);
  const extractTop = Math.floor(realCy - radius);

  const mask = Buffer.from(`<svg width="${cropSize}" height="${cropSize}"><circle cx="${cropSize/2}" cy="${cropSize/2}" r="${cropSize/2}" fill="#fff" /></svg>`);

  await sharp(imgPath)
    .extract({ left: extractLeft, top: extractTop, width: cropSize, height: cropSize })
    .ensureAlpha()
    .composite([{
      input: mask,
      blend: 'dest-in'
    }])
    .png()
    .toFile(destPng);
    
  console.log('Favicon created successfully. Extracted size: ' + cropSize + 'x' + cropSize);
}

makeFavicon().catch(console.error);
